feat(jans-auth-server): JIT compiler fails under jvm 17 (works perfectly under jvm 11)

### Description

JIT compiler fails under jvm 17 (works perfectly under jvm 11)
We need to find solution for jvm 17.

```
Caused by: java.lang.reflect.InaccessibleObjectException: Unable to make protected final java.lang.Class java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int) throws java.lang.ClassFormatError accessible: module java.base does not "opens java.lang" to unnamed module @77b52d12
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:354)
	at java.base/java.lang.reflect.AccessibleObject.checkCanSetAccessible(AccessibleObject.java:297)
	at java.base/java.lang.reflect.Method.checkCanSetAccessible(Method.java:199)
	at java.base/java.lang.reflect.Method.setAccessible(Method.java:193)
	at net.openhft.compiler.CompilerUtils.<clinit>(CompilerUtils.java:65)
```
Internal packages that will no longer be open by default: https://cr.openjdk.org/~mr/jigsaw/jdk8-packages-denied-by-default

--- a/jans-core/script/pom.xml
+++ b/jans-core/script/pom.xml
@@ -131,11 +131,6 @@
             <artifactId>testng</artifactId>
             <scope>test</scope>
         </dependency>
-
-        <dependency>
-            <groupId>net.openhft</groupId>
-            <artifactId>compiler</artifactId>
-        </dependency>
 	</dependencies>
 
 </project>

--- a/jans-core/script/src/main/java/io/jans/service/custom/script/ExternalTypeCreator.java
+++ b/jans-core/script/src/main/java/io/jans/service/custom/script/ExternalTypeCreator.java
@@ -8,14 +8,12 @@ import io.jans.model.custom.script.model.CustomScript;
 import io.jans.model.custom.script.model.ScriptError;
 import io.jans.model.custom.script.type.BaseExternalType;
 import io.jans.service.PythonService;
+import io.jans.service.custom.script.jit.SimpleJavaCompiler;
 import io.jans.util.StringHelper;
 import jakarta.enterprise.context.ApplicationScoped;
 import jakarta.inject.Inject;
-import net.openhft.compiler.CompilerUtils;
 import org.apache.commons.lang.StringUtils;
 import org.apache.commons.lang.exception.ExceptionUtils;
-import org.apache.logging.log4j.Level;
-import org.apache.logging.log4j.io.IoBuilder;
 import org.python.core.PyLong;
 import org.python.core.PyObject;
 import org.slf4j.Logger;
@@ -23,22 +21,16 @@ import org.slf4j.Logger;
 import java.io.ByteArrayInputStream;
 import java.io.IOException;
 import java.io.InputStream;
-import java.io.PrintWriter;
 import java.lang.reflect.InvocationTargetException;
-import java.net.URL;
-import java.net.URLClassLoader;
 import java.nio.charset.StandardCharsets;
 import java.util.Date;
 import java.util.Map;
-import java.util.concurrent.atomic.AtomicBoolean;
 
 /**
  * @author Yuriy Zabrovarnyy
  */
 @ApplicationScoped
 public class ExternalTypeCreator {
-
-    private static final AtomicBoolean classPathInitialized = new AtomicBoolean(false);
 
     @Inject
     protected Logger log;
@@ -48,22 +40,6 @@ public class ExternalTypeCreator {
 
     @Inject
     protected AbstractCustomScriptService customScriptService;
-
-    private void initJavaCompilerClasspathIfNeeded() {
-        if (classPathInitialized.get()) {
-            return;
-        }
-        try {
-            classPathInitialized.set(true);
-
-            URL[] urls = ((URLClassLoader) this.getClass().getClassLoader()).getURLs();
-            for (URL url : urls) {
-                CompilerUtils.addClassPath(url.getFile());
-            }
-        } catch (Throwable e) {
-            log.error("FAILED to output class loader urls", e);
-        }
-    }
 
     public BaseExternalType createExternalType(CustomScript customScript,
                                                Map<String, SimpleCustomProperty> configurationAttributes) {
@@ -81,6 +57,7 @@ public class ExternalTypeCreator {
         	loadException = ex;
             log.error("Failed to prepare external type '{}', exception: '{}'", customScriptInum, ExceptionUtils.getStackTrace(ex));
             log.error("Script '{}'", customScript.getScript());
+            log.error("Classpath '{}'", SimpleJavaCompiler.getClasspath());
         }
 
         externalType = initExternalType(externalType, customScript, configurationAttributes);
@@ -132,12 +109,8 @@ public class ExternalTypeCreator {
         return null;
     }
 
-    private BaseExternalType createExternalTypeWithJava(CustomScript customScript) throws ClassNotFoundException, NoSuchMethodException, IllegalAccessException, InvocationTargetException, InstantiationException {
-        initJavaCompilerClasspathIfNeeded();
-        CustomScriptType customScriptType = customScript.getScriptType();
-
-        PrintWriter printWriter = IoBuilder.forLogger(getClass()).setLevel(Level.DEBUG).buildPrintWriter();
-        Class<?> aClass = CompilerUtils.CACHED_COMPILER.loadFromJava(getClass().getClassLoader(), customScriptType.getClassName(), customScript.getScript(), printWriter);
+    private BaseExternalType createExternalTypeWithJava(CustomScript customScript) throws NoSuchMethodException, IllegalAccessException, InvocationTargetException, InstantiationException {
+        Class<?> aClass = SimpleJavaCompiler.compile(BaseExternalType.class, customScript.getScript());
         return (BaseExternalType) aClass.getDeclaredConstructor().newInstance();
     }
 

--- a/jans-core/script/src/main/java/io/jans/service/custom/script/jit/DiscardableClassLoader.java
+++ b/jans-core/script/src/main/java/io/jans/service/custom/script/jit/DiscardableClassLoader.java
@@ -1,0 +1,10 @@
+package io.jans.service.custom.script.jit;
+
+
+public class DiscardableClassLoader {
+    public static <T> Class<? extends T> classFromBytes(final Class<T> baseClass, final String name, final byte[] bytecode) {
+        return new ClassLoader(baseClass.getClassLoader()) {
+            Class<? extends T> c = defineClass(name, bytecode, 0, bytecode.length).asSubclass(baseClass);
+        }.c;
+    }
+}

--- a/jans-core/script/src/main/java/io/jans/service/custom/script/jit/IndentedPrintWriter.java
+++ b/jans-core/script/src/main/java/io/jans/service/custom/script/jit/IndentedPrintWriter.java
@@ -1,0 +1,136 @@
+package io.jans.service.custom.script.jit;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.Writer;
+
+
+public class IndentedPrintWriter extends PrintWriter {
+
+    /**
+     * Creates a new IndentedPrintWriter, without automatic line flushing.
+     *
+     * @param out A character-output stream
+     */
+    public IndentedPrintWriter(Writer out) {
+        super(new IndentedWriter(out));
+    }
+
+    /**
+     * Creates a new IndentedPrintWriter.
+     *
+     * @param out       A character-output stream
+     * @param autoFlush A boolean; if true, the <tt>println</tt>,
+     *                  <tt>printf</tt>, or <tt>format</tt> methods will
+     *                  flush the output buffer
+     */
+    public IndentedPrintWriter(Writer out, boolean autoFlush) {
+        super(new IndentedWriter(out), autoFlush);
+    }
+
+    /**
+     * Increases the indentation level
+     */
+    public void indent() {
+        ((IndentedWriter) out).indent();
+    }
+
+    /**
+     * Decreases the indentation level
+     */
+    public void dedent() {
+        ((IndentedWriter) out).dedent();
+    }
+
+    /**
+     * A writer that supports indentation of the output
+     */
+    @SuppressWarnings("WeakerAccess")
+    private static class IndentedWriter extends Writer {
+        /**
+         * Characters used for one indentation level
+         */
+        private static final char[] INDENT = "\t".toCharArray();
+        /**
+         * The underlying {@link Writer}
+         */
+        private final Writer out;
+        /**
+         * Indentation level
+         */
+        private int level;
+        /**
+         * true if must indent the next char
+         */
+        private boolean mustIndent;
+
+        /**
+         * Constructor based on a {@link Writer}
+         */
+        public IndentedWriter(Writer out) {
+            this.out = out;
+        }
+
+        /**
+         * Increases the indentation level
+         */
+        public void indent() {
+            ++level;
+        }
+
+        /**
+         * Decreases the indentation level
+         */
+        public void dedent() {
+            --level;
+        }
+
+        @Override
+        public void write(char[] cbuf, int off, int len) throws IOException {
+            int last = off;
+            for (int i = off; i < off + len; i++) {
+                if (mustIndent) {
+                    mustIndent = false;
+                    // Do not indent empty lines
+                    if (cbuf[i] != '\n')
+                        writeIndent();
+                }
+                if (cbuf[i] == '\n') {
+                    mustIndent = true;
+                    out.write(cbuf, last, i - last + 1);
+                    last = i + 1;
+                }
+            }
+            final int remaining = off + len - last;
+            if (remaining > 0)
+                out.write(cbuf, last, remaining);
+        }
+
+        private void writeIndent() throws IOException {
+            for (int i = 0; i < level; i++) {
+                out.write(INDENT);
+            }
+        }
+
+        @Override
+        public void flush() throws IOException {
+            out.flush();
+        }
+
+        @Override
+        public void close() throws IOException {
+            out.close();
+        }
+    }
+
+    /**
+     * Wraps the given writer in a indenting one if it isn't one already.
+     * Otherwise, returns the same one (casted).
+     */
+    public static IndentedPrintWriter of(PrintWriter writer) {
+        if (writer instanceof IndentedPrintWriter)
+            return (IndentedPrintWriter) writer;
+        else
+            return new IndentedPrintWriter(writer);
+    }
+}

--- a/jans-core/script/src/main/java/io/jans/service/custom/script/jit/JavaCodeGenerator.java
+++ b/jans-core/script/src/main/java/io/jans/service/custom/script/jit/JavaCodeGenerator.java
@@ -1,0 +1,557 @@
+package io.jans.service.custom.script.jit;
+
+import java.io.PrintWriter;
+import java.io.Writer;
+import java.util.*;
+import java.util.function.Consumer;
+
+
+public class JavaCodeGenerator {
+    /**
+     * Current scope
+     */
+    private Scope scope;
+
+    /**
+     * Parent of this code generator
+     */
+    private final JavaCodeGenerator parent;
+
+    /**
+     * Creates a new JavaCodeGenerator that writes to the specified writer
+     */
+    public JavaCodeGenerator() {
+        scope = new Scope();
+        parent = null;
+    }
+
+    /**
+     * Creates a new JavaCodeGenerator that generates in a sub-scope of the specified generator
+     */
+    private JavaCodeGenerator(JavaCodeGenerator parent) {
+        this.scope = new Scope(parent.scope);
+        this.parent = parent;
+    }
+
+    /**
+     * @return the name with a unique integer appended. The number is unique only for the current scope.
+     */
+    public String unique(String name) {
+        return scope.unique(name);
+    }
+
+    /**
+     * @return this code generator's parent or null if top-level
+     */
+    public JavaCodeGenerator parent() {
+        return parent;
+    }
+
+    /**
+     * Generates an import
+     */
+    public void generateImport(Class<?> aClass) {
+        printf("import %s;%n", aClass.getName());
+    }
+
+    /**
+     * Declares an uninitialized local variable
+     */
+    public JavaCodeGenerator declare(String type, String name) {
+        return declare(type, name, null);
+    }
+
+    /**
+     * Declare a pre-existing variable in the current scope. No code is emitted.
+     */
+    public JavaCodeGenerator declareExisting(String name) {
+        scope.declare(name);
+        return this;
+    }
+
+    /**
+     * Declares an initialized local variable
+     */
+    public JavaCodeGenerator declare(String type, String name, String initialValue, Object... args) {
+        scope.delayed.add(declareInternal(scope, type, name, initialValue == null ? null : String.format(initialValue, args)));
+        return this;
+    }
+
+    /**
+     * Declares an initialized local variable at the beginning of the current scope.
+     * Useful to declare temporary variables on complex expressions.
+     */
+    public JavaCodeGenerator declareOnScope(final String type, final String name, final String initialValue) {
+        scope.declare(type, name, initialValue);
+        return this;
+    }
+
+    /**
+     * Declares an initialized local variable at the beginning of the current scope.
+     * Useful to declare temporary variables on complex expressions.
+     *
+     * @return the resolved name of the variable
+     */
+    public static String declareOnScope(final Scope scope, final String type, final String name, final String initialValue) {
+        scope.delayed.add(scope.lastDeclaration++, declareInternal(scope, type, name, initialValue));
+        return scope.variable(name);
+    }
+
+    /**
+     * Prints the specified code at the beginning of the current scope.
+     * Useful to initialize temporary variables on complex expressions.
+     */
+    public JavaCodeGenerator printfOnScope(final String format, final Object... args) {
+        scope.delayed.add(scope.lastDeclaration++, printfInternal(format, args));
+        return this;
+    }
+
+    private Consumer<IndentedPrintWriter> printfInternal(final String format, final Object[] args) {
+        return pw -> pw.printf(format, args);
+    }
+
+    private static Consumer<IndentedPrintWriter> declareInternal(final Scope scope, final String type, String name, final String initialValue) {
+        scope.declare(name);
+        final String resolvedVariable = scope.variable(name);
+        return pw -> {
+            if (initialValue == null) {
+                pw.printf("%s %s;%n", type, resolvedVariable);
+            } else {
+                pw.printf("%s %s = %s;%n", type, resolvedVariable, initialValue);
+            }
+        };
+    }
+
+    /**
+     * Simple bounded loop over an int index variable
+     */
+    public JavaCodeGenerator boundedLoop(String indexVar, String lowerBound, String upperBound, Consumer<JavaCodeGenerator> body) {
+        enterScope();
+        scope.declare(indexVar);
+        printf("for (int %1$s = %2$s; %1$s < %3$s; %1$s++) {%n", variable(indexVar), lowerBound, upperBound);
+        indent();
+        body.accept(this);
+        dedent();
+        println("}");
+        leaveScope();
+        return this;
+    }
+
+    /**
+     * Unbounded loop
+     */
+    public JavaCodeGenerator loop(Consumer<JavaCodeGenerator> body) {
+        println("for (;;) {");
+        enterScope();
+        indent();
+        body.accept(this);
+        dedent();
+        println("}");
+        leaveScope();
+        return this;
+    }
+
+    /**
+     * Fore-each loop over an iterable
+     */
+    public JavaCodeGenerator forEach(String type, String elementVar, String collection, Consumer<JavaCodeGenerator> body) {
+        enterScope();
+        scope.declare(elementVar);
+        printf("for (%1$s %2$s : %3$s) {%n", type, variable(elementVar), collection);
+        indent();
+        body.accept(this);
+        dedent();
+        println("}");
+        leaveScope();
+        return this;
+    }
+
+    /**
+     * Emmit a single-line comment. Supports new lines.
+     */
+    public void comment(String s, Object... args) {
+        println("//" + String.format(s, args).replaceAll("\n", "\n//"));
+    }
+
+    /**
+     * @return the mangled name of the variable if it's shadowing another on an outer scope
+     */
+    public String variable(String name) {
+        return scope.variable(name);
+    }
+
+    /**
+     * Print the specified string so it is a valid java string literal.
+     */
+    public JavaCodeGenerator printQuoted(String s) {
+        if (s == null) {
+            print("null");
+        } else {
+            print('"');
+            print(escape(s));
+            print('"');
+        }
+        return this;
+    }
+
+    /**
+     * Generates a simple if-then statement
+     */
+    public JavaCodeGenerator ifThen(final String condition, Consumer<JavaCodeGenerator> accept) {
+        return ifThen(literal(condition), accept);
+    }
+
+    /**
+     * Generates a simple if-then statement
+     */
+    public JavaCodeGenerator ifThen(Consumer<JavaCodeGenerator> condition, Consumer<JavaCodeGenerator> accept) {
+        return ifElse(condition, accept, null);
+    }
+
+    /**
+     * Generate an if-else statement
+     */
+    public JavaCodeGenerator ifElse(String condition, Consumer<JavaCodeGenerator> accept, Consumer<JavaCodeGenerator> reject) {
+        return ifElse(literal(condition), accept, reject);
+    }
+
+    /**
+     * @return the current scope
+     */
+    public Scope getScope() {
+        return scope;
+    }
+
+    /**
+     * Generate an if-else statement
+     */
+    public JavaCodeGenerator ifElse(Consumer<JavaCodeGenerator> condition, Consumer<JavaCodeGenerator> accept, Consumer<JavaCodeGenerator> reject) {
+        print("if (");
+        condition.accept(this);
+        println(") {");
+        enterScope();
+        indent();
+        accept.accept(this);
+        dedent();
+        leaveScope();
+        if (reject != null) {
+            println("} else {");
+            enterScope();
+            indent();
+            reject.accept(this);
+            dedent();
+            leaveScope();
+        }
+        println("}");
+        return this;
+    }
+
+    private Consumer<JavaCodeGenerator> literal(final String condition) {
+        return cg -> cg.print(condition);
+    }
+
+    /**
+     * Close all open scopes and generate the code
+     */
+    public void generate(Writer out) {
+        final IndentedPrintWriter pw = new IndentedPrintWriter(out);
+        leaveUntil(parent == null ? null : parent.scope);
+        for (Consumer<IndentedPrintWriter> cmd : scope.delayed) {
+            cmd.accept(pw);
+        }
+    }
+
+    private void leaveUntil(Scope root) {
+        while (scope.parent != root)
+            leaveScope();
+    }
+
+    /**
+     * Begins a new block
+     */
+    public JavaCodeGenerator beginBlock() {
+        println("{");
+        indent();
+        enterScope();
+        return this;
+    }
+
+    /**
+     * Ends the current block
+     */
+    public JavaCodeGenerator endBlock() {
+        leaveScope();
+        dedent();
+        println("}");
+        return this;
+    }
+
+    private void enterScope() {
+        scope = new Scope(scope);
+    }
+
+    private void leaveScope() {
+        if (scope.parent == null) throw new IllegalStateException("you must be in a scope to leave it");
+        scope.parent.delayed.addAll(scope.delayed);
+        scope = scope.parent;
+    }
+
+    private String escape(String s) {
+        final int len = s.length();
+        StringBuilder sb = new StringBuilder(len);
+        for (int i = 0; i < len; i++) {
+            char c = s.charAt(i);
+            switch (c) {
+                case '"':
+                    sb.append("\\\"");
+                    break;
+                case '\\':
+                    sb.append("\\\\");
+                    break;
+                case '\n':
+                    sb.append("\\n");
+                    break;
+                case '\r':
+                    sb.append("\\r");
+                    break;
+                default:
+                    sb.append(c);
+                    break;
+            }
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Print a new line
+     */
+    public JavaCodeGenerator println() {
+        scope.delayed.add(PrintWriter::println);
+        return this;
+    }
+
+    /**
+     * Prints an Object and then terminates the line.
+     */
+    public JavaCodeGenerator println(final Object s) {
+        scope.delayed.add(pw -> pw.println(s));
+        return this;
+    }
+
+    /**
+     * Prints an object.
+     */
+    public JavaCodeGenerator print(final Object s) {
+        scope.delayed.add(pw -> pw.print(s));
+        return this;
+    }
+
+    /**
+     * Prints a formatted string using the specified format string and arguments.
+     */
+    public JavaCodeGenerator printf(final String format, final Object... args) {
+        scope.delayed.add(printfInternal(format, args));
+        return this;
+    }
+
+    private void indent() {
+        scope.delayed.add(IndentedPrintWriter::indent);
+    }
+
+    private void dedent() {
+        scope.delayed.add(IndentedPrintWriter::dedent);
+    }
+
+    /**
+     * Build a new public class
+     */
+    public ClassBuilder publicClass(String name) {
+        return new ClassBuilder("public", name);
+    }
+
+    /**
+     * Helper class to generate classes
+     */
+    public class ClassBuilder {
+        private ClassBuilder(String visibility, String name) {
+            declareExisting(name);
+            if (visibility != null) {
+                printf("%s class %s", visibility, name);
+            } else {
+                printf("class %s", name);
+            }
+        }
+
+        /**
+         * Make the generated class extend another class.
+         */
+        public ClassBuilder extend(String aClass) {
+            printf(" extends %s", aClass);
+            return this;
+        }
+
+        /**
+         * Make the generated class extend another class.
+         */
+        public ClassBuilder implement(String... anInterface) {
+            StringJoiner joiner = new StringJoiner(",");
+            for (String s : anInterface) {
+                joiner.add(s);
+            }
+            printf(" implements %s", joiner);
+            return this;
+        }
+
+        /**
+         * Build the class' body
+         */
+        public JavaCodeGenerator build(Consumer<JavaCodeGenerator> body) {
+            println(" {");
+            enterScope();
+            indent();
+            body.accept(JavaCodeGenerator.this);
+            dedent();
+            println("}");
+            leaveScope();
+            return JavaCodeGenerator.this;
+        }
+    }
+
+    /**
+     * Create a private method
+     */
+    public MethodBuilder packageMethod(String returnType, String name) {
+        return new MethodBuilder("", returnType, name);
+    }
+
+    /**
+     * Create a private method
+     */
+    public MethodBuilder privateMethod(String returnType, String name) {
+        return new MethodBuilder("private ", returnType, name);
+    }
+
+    /**
+     * Create a public method
+     */
+    public MethodBuilder publicMethod(String returnType, String name) {
+        return new MethodBuilder("public ", returnType, name);
+    }
+
+    /**
+     * Helper class to build methods
+     */
+    public class MethodBuilder {
+        private JavaCodeGenerator cg = new JavaCodeGenerator(JavaCodeGenerator.this);
+        private boolean hasArgs;
+
+        private MethodBuilder(String visibility, String returnType, String name) {
+            cg.printf("%s%s %s(", visibility, returnType, name);
+        }
+
+        /**
+         * Adds an argument of the specified type
+         */
+        public MethodBuilder arg(String type, String name) {
+            cg.declareExisting(name);
+            if (hasArgs) {
+                cg.print(", ");
+            }
+            cg.printf("%s %s", type, name);
+            hasArgs = true;
+            return this;
+        }
+
+        /**
+         * Builds this method's body
+         */
+        public JavaCodeGenerator build(Consumer<JavaCodeGenerator> body) {
+            cg.println(") {");
+            cg.indent();
+            // Generate the body in it's own scope.
+            final JavaCodeGenerator bodyGen = new JavaCodeGenerator(cg);
+            body.accept(bodyGen);
+            bodyGen.leaveScope();
+            cg.dedent();
+            cg.println("}");
+            cg.leaveScope();
+            return JavaCodeGenerator.this;
+        }
+    }
+
+    /**
+     * Simple scope for generation of shadowed local variables
+     */
+    public static class Scope {
+        private final Scope parent;
+        private final Set<String> variables = new HashSet<>();
+        private int lastDeclaration;
+        private final List<Consumer<IndentedPrintWriter>> delayed = new ArrayList<>();
+        private int varId;
+
+        private Scope() {
+            this.parent = null;
+        }
+
+        private Scope(Scope parent) {
+            this.parent = parent;
+        }
+
+        private void declare(String name) {
+            if (variables.contains(name))
+                throw new IllegalArgumentException(String.format("variable '%s' already declared in this scope", name));
+            variables.add(name);
+        }
+
+        private String mangle(String name) {
+            int level = -1;
+            for (Scope scope = this; scope != null; scope = scope.parent) {
+                if (scope.variables.contains(name)) {
+                    ++level;
+                }
+            }
+            return level > 0 ? name + "$" + level : name;
+        }
+
+        String variable(String name) {
+            if (variables.contains(name)) {
+                return mangle(name);
+            } else if (parent != null) {
+                return parent.variable(name);
+            } else {
+                throw new IllegalArgumentException("Unknown variable: " + name);
+            }
+        }
+
+        private String unique(String name) {
+            return name + "_" + (++varId);
+        }
+
+        /**
+         * Declares a variable at the beginning of the this scope.
+         */
+        public String declare(String type, String name, String initialValue) {
+            delayed.add(lastDeclaration++, declareInternal(type, name, initialValue));
+            return variable(name);
+        }
+
+        private Consumer<IndentedPrintWriter> declareInternal(final String type, String name, final String initialValue) {
+            declare(name);
+            final String resolvedVariable = variable(name);
+            return pw -> {
+                if (initialValue == null) {
+                    pw.printf("%s %s;%n", type, resolvedVariable);
+                } else {
+                    pw.printf("%s %s = %s;%n", type, resolvedVariable, initialValue);
+                }
+            };
+        }
+
+        /**
+         * @return true if the specified variable is defined in this scope
+         */
+        public boolean isDefined(String variable) {
+            return variables.contains(variable);
+        }
+    }
+}

--- a/jans-core/script/src/main/java/io/jans/service/custom/script/jit/SimpleJavaCompiler.java
+++ b/jans-core/script/src/main/java/io/jans/service/custom/script/jit/SimpleJavaCompiler.java
@@ -1,0 +1,226 @@
+package io.jans.service.custom.script.jit;
+
+import org.apache.commons.lang.StringUtils;
+
+import javax.tools.*;
+import java.io.*;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Consumer;
+
+/**
+ * @author Yuriy Z
+ */
+public class SimpleJavaCompiler {
+
+    /**
+     * A javac instance
+     */
+    private final JavaCompiler compiler;
+    /**
+     * The Standard file manager for the compiler
+     */
+    private final StandardJavaFileManager standardJavaFileManager;
+
+    /**
+     * We keep an initialized copy for each thread.
+     */
+    private static final ThreadLocal<SimpleJavaCompiler> BY_THREAD = new ThreadLocal<SimpleJavaCompiler>() {
+        @Override
+        protected SimpleJavaCompiler initialValue() {
+            return new SimpleJavaCompiler();
+        }
+    };
+
+    /**
+     * Create a new compiler
+     */
+    private SimpleJavaCompiler() {
+        compiler = ToolProvider.getSystemJavaCompiler();
+        standardJavaFileManager = compiler.getStandardFileManager(null, null, null);
+    }
+
+    /**
+     * Compiles and class-loads a single class from the specified Java source code.
+     * <p>
+     * To amortize the cost of initializing a JavaCompiler, underlying instances are cached
+     * on a per-thread basis. It is recommended that callers of this method reuse threads
+     * for increased performance and reduced memory consumption.
+     *
+     * @param source a String containing Java source code to generate the class
+     * @return the compiled and class-loaded class
+     */
+    public static <T> Class<? extends T> compile(Class<T> superClass, String source) {
+        return BY_THREAD.get().compile0(superClass, source);
+    }
+
+
+    /**
+     * Compiles and class-loads a single class from the specified Java source code.
+     * <p>
+     * To amortize the cost of initializing a JavaCompiler, underlying instances are cached
+     * on a per-thread basis. It is recommended that callers of this method reuse threads
+     * for increased performance and reduced memory consumption.
+     *
+     * @param sourceGen a String containing Java source code to generate the class
+     * @return the compiled and class-loaded class
+     */
+    public static <T> Class<? extends T> compile(Class<T> superClass, Consumer<JavaCodeGenerator> sourceGen) {
+        final JavaCodeGenerator cg = new JavaCodeGenerator();
+        sourceGen.accept(cg);
+
+        final StringWriter sw = new StringWriter();
+        cg.generate(sw);
+
+        final Class<? extends T> compiledClass = compile(superClass, sw.toString());
+        dump(sw, compiledClass);
+        return compiledClass;
+    }
+
+    public static volatile boolean dump = Boolean.getBoolean("compiler.java.dump");
+
+    public static volatile String dumpDir = System.getProperty("compiler.java.dump.dir");
+
+    private static void dump(StringWriter sw, Class<?> compiledClass) {
+        if (dump) {
+            System.out.println(sw);
+        }
+        if (dumpDir != null) {
+            try {
+                Files.write(Paths.get(dumpDir, compiledClass.getCanonicalName().replace('.', File.separatorChar) + ".java"), sw.toString().getBytes());
+            } catch (IOException e) {
+                throw new RuntimeException("unable to dump source file", e);
+            }
+        }
+    }
+
+    private static String classpath;
+
+    public static String getClasspath() {
+        if (StringUtils.isBlank(classpath)) {
+            URL[] urls = ((URLClassLoader) SimpleJavaCompiler.class.getClassLoader()).getURLs();
+
+            StringBuilder sb = new StringBuilder(System.getProperty("java.class.path"));
+            sb.append(File.pathSeparator);
+            for (URL url : urls) {
+                sb.append(url.toString()).append(File.pathSeparator);
+            }
+
+            classpath = sb.toString();
+
+        }
+        return classpath;
+    }
+
+    /**
+     * Compiles and class-loads a single class from the specified Java source code.
+     *
+     * @param source a String containing Java source code to generate the class
+     * @return the compiled and class-loaded class
+     */
+    private <T> Class<? extends T> compile0(Class<T> superClass, String source) {
+        final List<JavaFileObject> compilationUnits = Collections.singletonList(new SourceFile(toUri("Generated"), source));
+        final FileManager fileManager = new FileManager(standardJavaFileManager);
+        final StringWriter output = new StringWriter();
+        final JavaCompiler.CompilationTask task = compiler.getTask(output, fileManager, null, Arrays.asList("-g", "-proc:none", "-classpath", getClasspath()), null, compilationUnits);
+
+        if (!task.call()) {
+            throw new IllegalArgumentException("Compilation failed:\n" + output + "\n Source code: \n" + source);
+        }
+
+        int classCount = fileManager.output.size();
+        // 2nd condition to work around the case where additional "GuardedBy" generated in TA Spark tagging
+        if (classCount == 1 || (classCount > 0 && "CompiledStage"
+                .equals(fileManager.output.get(0).getName()))) {
+            return DiscardableClassLoader.classFromBytes(superClass, null, fileManager.output.get(0).outputStream.toByteArray());
+        }
+        throw new IllegalArgumentException("Compilation yielded an unexpected number of classes: " + classCount);
+    }
+
+    /**
+     * Represents a source file whose contents is loaded from a String
+     */
+    private static class SourceFile extends SimpleJavaFileObject {
+        private final String contents;
+
+        private SourceFile(URI uri, String contents) {
+            super(uri, Kind.SOURCE);
+            this.contents = contents;
+        }
+
+        @Override
+        public String getName() {
+            return uri.getRawSchemeSpecificPart();
+        }
+
+        /**
+         * Ignore the file name for public classes
+         */
+        @Override
+        public boolean isNameCompatible(String simpleName, Kind kind) {
+            return true;
+        }
+
+        @Override
+        public CharSequence getCharContent(boolean ignoreEncodingErrors) throws IOException {
+            return contents;
+        }
+    }
+
+    /**
+     * A compiled class file. It's content is stored in a ByteArrayOutputStream.
+     */
+    private static class ClassFile extends SimpleJavaFileObject {
+        private ByteArrayOutputStream outputStream;
+
+        private ClassFile(URI uri) {
+            super(uri, Kind.CLASS);
+        }
+
+        @Override
+        public String getName() {
+            return uri.getRawSchemeSpecificPart();
+        }
+
+        @Override
+        public OutputStream openOutputStream() throws IOException {
+            return outputStream = new ByteArrayOutputStream();
+        }
+    }
+
+    /**
+     * A simple file manager that collects written files in memory
+     */
+    private static class FileManager
+            extends ForwardingJavaFileManager<StandardJavaFileManager> {
+
+        private List<ClassFile> output = new ArrayList<>();
+
+        FileManager(StandardJavaFileManager target) {
+            super(target);
+        }
+
+        @Override
+        public JavaFileObject getJavaFileForOutput(Location location, String className, JavaFileObject.Kind kind, FileObject sibling) {
+            final ClassFile file = new ClassFile(toUri(className));
+            output.add(file);
+            return file;
+        }
+    }
+
+    private static URI toUri(String path) {
+        try {
+            return new URI(null, null, path, null);
+        } catch (URISyntaxException e) {
+            throw new RuntimeException("exception parsing uri", e);
+        }
+    }
+}

--- a/jans-core/script/src/test/java/io/jans/service/custom/script/test/java/SimpleJavaCompileTest.java
+++ b/jans-core/script/src/test/java/io/jans/service/custom/script/test/java/SimpleJavaCompileTest.java
@@ -1,24 +1,19 @@
 package io.jans.service.custom.script.test.java;
 
-import net.openhft.compiler.CompilerUtils;
-
-import java.lang.reflect.InvocationTargetException;
+import io.jans.service.custom.script.jit.SimpleJavaCompiler;
 
 /**
  * @author Yuriy Zabrovarnyy
  */
 public class SimpleJavaCompileTest {
 
-    public static void main(String[] args) throws ClassNotFoundException, IllegalAccessException, InstantiationException, NoSuchMethodException, InvocationTargetException {
-        String className = "MyClass";
-        String javaCode =
-                "public class MyClass implements Runnable {\n" +
-                "    public void run() {\n" +
-                "        System.out.println(\"Hello World\");\n" +
-                "    }\n" +
-                "}\n";
-        Class aClass = CompilerUtils.CACHED_COMPILER.loadFromJava(className, javaCode);
-        Runnable runner = (Runnable) aClass.getDeclaredConstructor().newInstance();
-        runner.run();
+    public static void main(String[] args) throws Exception {
+        String source = "import io.jans.service.custom.script.test.java.Testable; public class Test implements Testable { static { System.out.println(\"hello\"); } public Test() { System.out.println(\"world\"); } }";
+
+        for (int i = 0; i < 10000; i++) {
+            Class<Testable> cls = (Class<Testable>) SimpleJavaCompiler.compile(Testable.class, source);
+            Object instance = cls.getDeclaredConstructor().newInstance();
+            System.out.println(i);
+        }
     }
 }

--- a/jans-core/script/src/test/java/io/jans/service/custom/script/test/java/Testable.java
+++ b/jans-core/script/src/test/java/io/jans/service/custom/script/test/java/Testable.java
@@ -1,0 +1,7 @@
+package io.jans.service.custom.script.test.java;
+
+/**
+ * @author Yuriy Z
+ */
+public interface Testable {
+}

